_posts/2015-12-01-minutes-offline-case-study.md

```diff
@@ -0,0 +1,101 @@
+
+## Handling tasks that cannot possibly be completed without a network connection
+
+Some things can’t be done offline. We can’t send emails without a network connection, period. We can’t do things on the server if we can’t reach it. To help us handle cases like this, Hoodie’s task API lets us define tasks we want the server to perform, and syncs them like any other piece of data. So the task __”hey server, please email this meeting to the following addresses”__ is stored locally in the browser, and synced to the server whenever possible, which will then send the emails and sync back a task confirmation whenever possible. When the user is online, this simply sends the emails immediately, because the process is practically instantaneous, but when they’re offline, nothing breaks. It’s like any email client’s outbox.
+
+## Handling image uploads when all you’ve got is localStorage and no internet
+
+I wasn’t being entirely truthful earlier when I said that minutes.io users only store text data: they can also add custom logos to their meeting notes and emails. Of course, since we’re currently limited to localStorage, we can’t store huge images. Also, since this feature has to work offline, we can’t rely on the server to do anything for us, like image resizing. Luckily, neither of these issues is a problem.
+
+__We don’t need huge images:__ the logos are displayed in a 126x126px square, so even with 2x retina images, that’s not much data. 2x is also pretty fine for printing. Plus, we don’t need to let people upload infinite numbers of logos: for now, we let users have a small library of three to pick from. And finally, instead of actually attaching the logo image to every individual meeting, meetings only reference one of the three possible images in a user’s logo library, which keeps data size minimal.
+
+__It’s possible to do all image ”uploading”, manipulation and storage in-browser.__ We can grab the image data from the File object without actually uploading anything, we can use canvas with [limby-resize](https://github.com/danschumann/limby-resize) to get crisply shrunk images, and we use canvas’ __toDataURL(’image/jpeg’, 0.8)__ to output a string version of the nicely compressed logos. The stringified logo image is then just saved via __hoodie.store.add()__, like any other meeting data, and synced in the same way.
+
+The result is a feature that just works, online or offline, and is completely opaque to the user. There’s no difference in functionality or behaviour when offline, neither here or at any other point in the app. Users can ”upload” logos when offline, just like they can save meetings or click the ”Send email” button when offline. The app just deals with it and communicates accordingly.
```

Comment:
A lot of "just" repetition. I suggested "app just deals with it" > "app deals with it" to reduce repetition.